cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+        }
+        
+        CloseDB(db_cached);
+        return have_error ? -1 : 0;
+    }
+    /* Unable to open database. */
+    return -1;
+}
+
+
+bool UpdateCache(Rlist* options, const PackageModuleWrapper *wrapper,
+                 UpdateType type)
+{
+    char *options_str = ParseOptions(options);
+    Rlist *response = NULL;
+    
+    const char *req_type = NULL;
+    if (type == UPDATE_TYPE_INSTALLED)
+    {
+        req_type = "list-installed";
+    }
+    else if (type == UPDATE_TYPE_UPDATES)
+    {
+        req_type = "list-updates";
+    }
+    else if (type == UPDATE_TYPE_LOCAL_UPDATES)
+    {
+        req_type = "list-updates-local";
+    }
+    if (ReadWriteDataToPackageScript(req_type, options_str, &response,
+            wrapper) != 0)
+    {
+        Log(LOG_LEVEL_VERBOSE, "Some error occurred while communicating with "
+                "package module while updating cache.");
+        free(options_str);
+        return false;
+    }
+    
+    if (!response)
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Received empty packages list after requesting: %s", req_type);
+    }
+    
+    bool ret = true;
+    
+    /* We still need to update DB with empty data. */
+    if (UpdatePackagesDB(response, wrapper->name, type) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Error updating packages cache.");
+        ret = false;
+    }
+    
+    RlistDestroy(response);
+    free(options_str);
+    return ret;
+}
+
+
+PromiseResult ValidateChangedPackage(const NewPackages *policy_data,
+                                     const PackageModuleWrapper *wrapper,
+                                     const PackageInfo *package_info,
+                                     NewPackageAction action_type)
+{
+    
+    if (!UpdateCache(policy_data->package_options, wrapper,
+                     UPDATE_TYPE_INSTALLED))
+    {
+        Log(LOG_LEVEL_INFO, "Can not update installed packages cache after "
+            "package installation");
+        return PROMISE_RESULT_FAIL;
+    }
+    
+    if (!UpdateCache(policy_data->package_options, wrapper,
+                     UPDATE_TYPE_LOCAL_UPDATES))
+    {
+        Log(LOG_LEVEL_INFO, "Can not update available updates cache after "
+            "package installation");
+        return PROMISE_RESULT_FAIL;
+    }
+
+    int is_in_cache = IsPackageInCache(NULL, wrapper, package_info->name,
+                                       package_info->version,
+                                       package_info->arch);
+    if (is_in_cache == 1)
+    {
+        return action_type == NEW_PACKAGE_ACTION_PRESENT ? 
+            PROMISE_RESULT_CHANGE : PROMISE_RESULT_FAIL;
+    }
+    else if (is_in_cache == 0)
+    {
+        return action_type == NEW_PACKAGE_ACTION_PRESENT ? 
+            PROMISE_RESULT_FAIL : PROMISE_RESULT_CHANGE;
+    }
+    else
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while reading installed "
+            "packages cache.");
+        return PROMISE_RESULT_FAIL;
```

Comment:
All this function's Log()s could line-break after the level and rejoin the two halves of their messages.